### PR TITLE
Do not put the closing quotes in a docstring on a separate line

### LINKED ISSUE #3320: style regression: too long single line docstrings have their quotes moved to a new line.
**Describe the bug**

Single-line docstrings that are longer than the line length have the final three `"` moved to a new line.

**To Reproduce**

For example, take this code:
```py
from typing import Optional

# leaves this the same
def large_image_url(self) -> Optional[str]:
    """Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable.
    """
    pass


# leaves this the same too
def large_image_url(self) -> Optional[str]:
    """
    Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable.
    """
    pass


# depending on the length of the docstring, this one *may* be left as-is
def large_image_url(self) -> Optional[str]:
    """
    Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable."""
    pass


# but this one is modified
def large_image_url(self) -> Optional[str]:
    """Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable."""
    pass

```

And run it with these arguments:

```sh
$ black file.py --preview
```

We get this format:

```py
from typing import Optional

# leaves this the same
def large_image_url(self) -> Optional[str]:
    """Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable.
    """
    pass


# leaves this the same too
def large_image_url(self) -> Optional[str]:
    """
    Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable.
    """
    pass


# this one is also left as-is
def large_image_url(self) -> Optional[str]:
    """
    Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable.
    """
    pass


# but this one is modified
def large_image_url(self) -> Optional[str]:
    """Optional[:class:`str`]: Returns a URL pointing to the large image asset of this activity, if applicable.
    """
    pass
```

**Expected behavior**

Single line docstrings are left alone. I understand what this change is supposed to do, and on already multiline docstrings this wouldn't be an issue (although it would be nice for the docstring formatter to be entirely consistent among all docstrings instead of having different behavior for the line length but i digress).

**Environment**

- Black's version:  22.10.0
- OS and Python version:  Python 3.11.0rc2, on linux, but i can recreate with python 3.8 too

**Additional Comments**

This seems like a regression from #3044

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Fixes #3320. Followup from #3044.

### CODE REVIEW

**Review by ichard26 — COMMENTED** [review]
Code looks fine in general, but this code is hard to grok so let's see what diff-shades has to say first.

**Comment by JelleZijlstra** [comment]
The diff-shades report is as expected, just removing newlines between one-line docstrings and the closing quotes. Example:

```
 def _static_hasattr(value: object, attr: str) -> bool:
-    """Returns whether this value has the given attribute, ignoring __getattr__ overrides.
-    """
+    """Returns whether this value has the given attribute, ignoring __getattr__ overrides."""
     try:
```

File: src/black/linegen.py [hunk 1]
```diff
@@ -389,19 +389,18 @@ def visit_STRING(self, leaf: Leaf) -> Iterator[Line]:
...
-                    last_line_length += len(indent) + len(prefix) + quote_len
-
                 # If adding closing quotes would cause the last line to exceed
                 # the maximum line length then put a line break before the
                 # closing quotes
-                if last_line_length + quote_len > self.mode.line_length:
+                if (
+                    len(lines) > 1
+                    and last_line_length + quote_len > self.mode.line_length
+                ):
```
Comment (ichard26) [inline src/black/linegen.py:403]: Why does this not take indentation into account when determining whether the line length would be violated?

**Review by ichard26 — COMMENTED** [review]
On second look, this code looks suspicious.

File: src/black/linegen.py (hunk 1 above)
Comment (ichard26) [reply src/black/linegen.py:403]: Oh hmm, sometimes it does take into account the indentation. Looks like `lines` doesn't contain what I expected it to. Nothing here makes sense :)

**Review by ichard26 — APPROVED** [review]
Actually this seems fine on a third look.

**Comment by ichard26** [comment]
Thanks!

**Comment by jensguballa** [comment]
I have created a [comment](https://github.com/psf/black/pull/3044#issuecomment-1366026400) on #3044, but as this here is the issue that has been merged, I am providing my thought here again.

After going over the issue I am a little bit puzzled now, as I am not really sure what the intended behavior should be.

My assumption was that black should not apply reformattings which will introduce new line length violations. In my opinion this sounds reasonable, given that you might have a processing pipeline where e.g. flake8 is doing its job hereafter and would fail (like in my case). But it looks like the resolution for this issue is going in another direction.

My view: For the [example](https://github.com/psf/black/pull/3044#issuecomment-1364710654) I have provided black needs to take a decision: either reformatting to a one-line string and violating the maximum line length, or, leaving the doc string as it is and keeping the conformity.

My preference is clearly to keep the conformity, but if I understand the resolution here correctly, reformatting to a one-line string is given priority. I am not in favor of this as special crafting of the doc string by the developer just to work around this issue would go against the philosophy of black.

So would it be reasonable not to reformat the doc string for the example I provided?

**Comment by ichard26** [comment]
The crux of the problem here is that Black shouldn't move quotes that already violate the line length limit onto a new line (it's ugly and isn't compliant with PEP 257), but _also_ shouldn't move quotes that were already on their own line. This is impossible to do w/ Black's philosophy of avoiding looking at the original formatting as much as possible.  

I agree Black shouldn't cause more violations of the line length limit. Technically this introduces one more situation where Black is dependant on pre-existing formatting, but I don't think docstrings can be formatted in an "one size for all" way IMHO.

**Comment by jensguballa** [comment]
I do know nothing about black's internals, but for the example I provided would it be possible to implement something like this for one-line doc strings:

Original code (maximum line length not exceeded):
```
class _YamlBlockStyle(str):
    """Class used to indicate that a string shall be serialized using the block style.
    """
```

What black produces currently (one-line doc string, maximum line length exceeded):
```
class _YamlBlockStyle(str):
    """Class used to indicate that a string shall be serialized using the block style."""
```

What I propose black could produce (multi-line doc string, maximum line length not exceeded):
```
class _YamlBlockStyle(str):
    """
    Class used to indicate that a string shall be serialized using the block style.
    """
```

I think that output would comply to the maximum line length as well as to PEP 257. If the latter output would still exceed the maximum line length, then of course there is nothing black can do.

**Comment by adamjstewart** [comment]
If I'm understanding correctly, this PR re-introduces the following issues: #1632, #2274, #3331?

It's great that we're now PEP 257 compliant, but now black reformats my code such that flake8 complains about line length. What's the suggested approach for users using both black and flake8, to add `# noqa: E501` to every docstring in this situation? I would argue that this is uglier than having the closing triple quotes on a separate line...

**Pull request merged**